### DISCUSSION TITLE
New package for writing Postgres functions in Typescript

### DISCUSSION THREAD

**Discussion by davepar** [supabase/supabase#11380] [Show and tell]
I decided to create some support for writing PostgreSQL PLV8 functions in Typescript. The result is an [NPM package](https://www.npmjs.com/package/plv8ts) and [an article](https://dev.to/davepar/postgresql-functions-in-typescript-27dk) with more background. Instead of having Javascript buried in a `CREATE FUNCTION` SQL statement, you can write Typescript with editor support, compile-time syntax checking, and unit tests.

Feedback welcome. This is my first NPM package.

**Comment by pkit** [comment]
Looks nice!
Lacks some features though.
From the top of my head, as somebody who wrote a closed source implementation of these ~4 years ago:
1. Simple stuff like `search` path, `schema` and `grant` are tremendously useful.
2. Bundling the code, writing everything in one function is not scalable... 
Also basic `require()` can be added for the "core" modules. Imports can be transpiled into `require()` anyway.
3. Wrappers, derives from the point above. 
Error stacks become impossible to unwind, so some `*.map` is needed, and support for unpacking it.
4. Polyfills, like `crypto`, used a lot in nodejs. Can be trivially supported with `pg_crypto`, etc. 
Base packages can be lifted from `browserify` and `rollup` bundles.
5. `setTimeout` support through `pg_net`-like interface with an edge function for resume. (it really works, I've tried).
6. Promise support, by leveraging `setTimeout`
7. Filesystem support, a lot of packages use `fs` module without thinking twice. Can be implemented with postgres large objects + `browserfs` + `ltree` extension.
8. Full HTTP/1.1 API GW, can be done on top of other features above. Mine used the `express.js` codebase.

**Comment by james-pre** [comment]
Hi @pkit, I'm the maintainer of BrowserFS. I'd like to let you know that BrowserFS has been deprecated. You can read more about that on the BrowserFS readme (https://github.com/jvilk/BrowserFS). [ZenFS](https://github.com/zen-fs/core) is an updated and active fork that I created.
